### ISSUE TITLE
Android/iOS Accessible Live Regions

### ISSUE THREAD

**Issue by aliossam** [facebook/react-native#34735]
### Description

The accessibilityLiveRegion prop is currently only supported on Android, and there are issues with it even on that platform. Ideally this prop would consistently work on both iOS and Android.

**Expected Behavior** 
Whenever an element with accessibilityLiveRegion set, or any descendant of an element with accessibilityLiveRegion set has any content changes, they should be queued up to be announced by the screen reader. “Polite” announcements should not interrupt any ongoing announcements, while “assertive” ones should interrupt.

**Android Details** 
On Android, the accessibilityLiveRegion prop works in the older Paper renderer, but not the newer Fabric renderer. 

**iOS Details** 

On iOS there is no concept of live regions at a system level, so to support this we’d need to detect when content changes on any element (or descendant) with the accessibilityLiveRegion prop set, and make a manual announcement. These announcements will have to be queued up with the UIAccessibilitySpeechAttributeQueueAnnouncement property set to either make them “polite” or “assertive”. 

The behavior of exactly how the queue works (maximum length, what happens when new announcements are queued after that length, etc.) should match Android and web as closely as possible.

### Version

0.66

### Output of `npx react-native info`

see description

### Steps to reproduce

see description

### Snack, code example, screenshot, or link to a repository

see description

**Comment by fabriziobertoglio1987** [comment]
The functionality requires creation of a queue for iOS announcements. While polite seems to be possible to implement, assertive may be complex. It requires to interrupt a system announcement. 

- [ ] Review information in Issue [#34735](https://github.com/facebook/react-native/issues/34735)
- [ ] Read about iOS API UIAccessibilitySpeechAttributeQueueAnnouncement ([stackoverflow](https://stackoverflow.com/a/52459406/7295772))
Read iOS documentation on [UIAccessibilitySpeechAttributeQueueAnnouncement](https://developer.apple.com/documentation/uikit/uiaccessibilityspeechattributequeueannouncement?language=objc)
- [ ] Test existing reactnative accessibilityLiveRegion functionality on Android Paper
- [ ] Test [UIAccessibilitySpeechAttributeQueueAnnouncement](https://stackoverflow.com/a/67951073/7295772) in react-native
Search for the APIs discussed on [stackoverflow](https://stackoverflow.com/a/52459406/7295772) answer in [AccessibilityAnnouncer](https://github.com/spanage/AccessibilityAnnouncer) library"

**Comment by fabriziobertoglio1987** [comment]
#### Further investigate [announcementFinished](https://github.com/facebook/react-native/commit/cfe003238ab8c5686d185f6ce9e0776eeb4bb729) API

https://github.com/facebook/react-native/blob/cfe003238ab8c5686d185f6ce9e0776eeb4bb729/Libraries/Components/AccessibilityInfo/AccessibilityInfo.ios.js#L95-L124

**Comment by fabriziobertoglio1987** [comment]
#### Test [UIAccessibilitySpeechAttributeQueueAnnouncement](https://stackoverflow.com/a/67951073/7295772) in react-native ([StackOverflow](https://stackoverflow.com/q/50856959/7295772))

https://reactnative.dev/docs/next/accessibilityinfo#announceforaccessibilitywithoptions

<details><summary>RCTAccessibilityManager</summary>
<p>

```objc
RCT_EXPORT_METHOD(announceForAccessibilityWithOptions
                  : (NSString *)announcement options
                  : (JS::NativeAccessibilityManager::SpecAnnounceForAccessibilityWithOptionsOptions &)options)
{
  if (@available(iOS 11.0, *)) {
    NSMutableDictionary<NSString *, NSNumber *> *attrsDictionary = [NSMutableDictionary new];
    if (options.queue()) {
      attrsDictionary[UIAccessibilitySpeechAttributeQueueAnnouncement] = @(*(options.queue()) ? YES : NO);
    }

    if (attrsDictionary.count > 0) {
      NSAttributedString *announcementWithAttrs = [[NSAttributedString alloc] initWithString:announcement
                                                                                  attributes:attrsDictionary];
      UIAccessibilityPostNotification(UIAccessibilityAnnouncementNotification, announcementWithAttrs);
    } else {
      UIAccessibilityPostNotification(UIAccessibilityAnnouncementNotification, announcement);
    }
  } else {
    UIAccessibilityPostNotification(UIAccessibilityAnnouncementNotification, announcement);
  }
}
```

</p>
</details>

**Comment by fabriziobertoglio1987** [comment]
<details><summary>Testing Android accessibilityLiveRegion assertive (Paper)</summary>
<p>

```jsx
        <RNTesterBlock title="LiveRegion">
          <TouchableWithoutFeedback onPress={this._addOne}>
            <View style={styles.embedded}>
              <Text>Click me</Text>
            </View>
          </TouchableWithoutFeedback>
          <View accessibilityLiveRegion="polite">
            <Text>
              Announced {this.state.count} times with accessibilityLiveRegion
              polite
            </Text>
          </View>
          <View accessibilityLiveRegion="assertive">
            <Text>
              Second announcement {this.state.count} times with
              accessibilityLiveRegion assertive
            </Text>
          </View>
        </RNTesterBlock>
```

https://www.icloud.com/iclouddrive/08b3HMFKoLEfpMCCyX9enjyIw#android_paper_assertive

</p>
</details>

<details><summary>Testing Android accessibilityLiveRegion polite (Paper)</summary>
<p>

```jsx
        <RNTesterBlock title="LiveRegion">
          <TouchableWithoutFeedback onPress={this._addOne}>
            <View style={styles.embedded}>
              <Text>Click me</Text>
            </View>
          </TouchableWithoutFeedback>
          <View accessibilityLiveRegion="polite">
            <Text>
              Announced {this.state.count} times with accessibilityLiveRegion
              polite
            </Text>
          </View>
          <View accessibilityLiveRegion="polite">
            <Text>
              Second announcement {this.state.count} times with
              accessibilityLiveRegion polite
            </Text>
          </View>
        </RNTesterBlock>
```

https://www.icloud.com/iclouddrive/078YVxbmyt9XswScPrBtTuGrg#android_paper_polite

</p>
</details>

**Comment by fabriziobertoglio1987** [comment]
## Design Implementation

1) Different react components have the accessibilityLiveRegion=polite or assertive
```jsx
<>
  <TouchableWithoutFeedback onPress={this._addOne}>
    <View style={styles.embedded}>
      <Text>Click me</Text>
    </View>
  </TouchableWithoutFeedback>
  <View accessibilityLiveRegion="polite">
    <Text>
      Announced {this.state.count} times with accessibilityLiveRegion
      polite
    </Text>
  </View>
  <View accessibilityLiveRegion="polite">
    <Text>
      Second announcement {this.state.count} times with
      accessibilityLiveRegion polite
    </Text>
  </View>
</>
```
2) If their content changes, an announcement is triggered:
- [ ] Find API AccessibilityNodeInfo onContentChange/onNodeChange (for iOS)
- [ ] Trigger manual announcement from the onContentChange callback
     
3) The above announcement should be assertive or polite (more info in [comment](https://github.com/facebook/react-native/issues/34735#issuecomment-1322274507))

**Comment by fabriziobertoglio1987** [comment]
#### If their content changes, an announcement is triggered:
- [ ] Verify conditions that trigger the announcement for content change (accessibilityLabel, accessibilityHint, and other component props)
- [ ] Debug the Text.js react-native component (lifecycle, state) to detect a content change
       Trigger VoiceOver announcement onContentChange.
- [ ] Find API AccessibilityNodeInfo onContentChange/onNodeChange (for iOS)
- [ ] Trigger manual announcement from the onContentChange callback

**Comment by fabriziobertoglio1987** [comment]
## Automatic content grouping

Solution proposted in PR https://github.com/facebook/react-native/pull/35432 does not work with automatic content grouping (for ex. child component Text change does not trigger a new announcement when focused on another component).

<details><summary>CLICK TO OPEN TESTS RESULTS</summary>
<p>

```javascript
function AccessibilityExpandedExample(): React.Node {
  const [expand, setExpanded] = React.useState(false);
  const [pressed, setPressed] = React.useState(false);
  const expandAction = {name: 'expand'};
  const collapseAction = {name: 'collapse'};
  return (
    <>
      <RNTesterBlock title="Collapse/Expanded state change (Paper)">
        <Button
          onPress={() => {
            setExpanded(!expand);
          }}
          title="click me to change state"
        />
        <View
          accessibilityLiveRegion="polite"
          accessibilityRole="button"
          accessible={true}
          focusable={true}
          nativeID={'1'}
          style={{backgroundColor: 'red', height: 200, width: 400}}>
          <Text>{expand ? null : 'my text'}</Text>
        </View>
      </RNTesterBlock>
    </>
  );
}
```


</p>
</details>

The child text component attributedString.text is copied to the parent accessibilityLabel, but only when VoiceOver focus moves to the component. 

https://github.com/facebook/react-native/blob/53f7b76ca885d7a50af17b34b87161e1f72b5ee2/React/Fabric/Mounting/ComponentViews/View/RCTViewComponentView.mm#L701-L726

https://github.com/facebook/react-native/blob/53f7b76ca885d7a50af17b34b87161e1f72b5ee2/React/Fabric/Mounting/ComponentViews/Text/RCTParagraphComponentView.mm#L138-L141

While the prop diffing is triggered when the new prop is passed to the child component, as there is no change in the accessibilityLabel prop, no announcement is triggered here.

https://github.com/facebook/react-native/blob/53f7b76ca885d7a50af17b34b87161e1f72b5ee2/React/Fabric/Mounting/ComponentViews/View/RCTViewComponentView.mm#L298-L308

**Comment by fabriziobertoglio1987** [comment]
https://github.com/facebook/react-native/blob/53f7b76ca885d7a50af17b34b87161e1f72b5ee2/ReactCommon/react/renderer/components/text/ParagraphShadowNode.h#L94-L98

https://github.com/facebook/react-native/blob/53f7b76ca885d7a50af17b34b87161e1f72b5ee2/ReactCommon/react/renderer/components/text/ParagraphShadowNode.cpp#L100-L115

https://github.com/facebook/react-native/blob/53f7b76ca885d7a50af17b34b87161e1f72b5ee2/React/Fabric/Mounting/ComponentViews/Text/RCTParagraphComponentView.mm#L101-L106

https://github.com/facebook/react-native/blob/53f7b76ca885d7a50af17b34b87161e1f72b5ee2/React/Fabric/Mounting/ComponentViews/Text/RCTParagraphComponentView.mm#L60-L64

https://github.com/facebook/react-native/blob/53f7b76ca885d7a50af17b34b87161e1f72b5ee2/React/Fabric/Mounting/ComponentViews/Text/RCTParagraphComponentView.mm#L138-L141

**Comment by fabriziobertoglio1987** [comment]
<details><summary>verify the value passed through updateState</summary>
<p>

https://github.com/facebook/react-native/blob/53f7b76ca885d7a50af17b34b87161e1f72b5ee2/React/Fabric/Mounting/RCTMountingManager.mm#L133

the previous text is empty, new text value is "my text"
<img width="1920" alt="state_update_child_text_to_null" src="https://user-images.githubusercontent.com/24992535/203631085-d2320d6d-9bab-4efd-a1c8-4fd9a85e0072.png">

the previous text is "my text", new text value is empty
<img width="1920" alt="state_update_child_text_to_my_text" src="https://user-images.githubusercontent.com/24992535/203631109-0197fa76-2a0e-4007-a6c2-1dd64f84a5d3.png">

</p>
</details>

**Comment by fabriziobertoglio1987** [comment]
<details><summary>Android - testing announcement of accessibilityState selected with accessibilityLiveRegion</summary>
<p>

https://user-images.githubusercontent.com/24992535/204022975-7f5361f4-e587-4fef-be90-0dfeb8bd31e9.mp4

</p>
</details>

**Comment by fabriziobertoglio1987** [comment]
This is an iOS issue in the main branch, the Text accessibilityLabel is set to always equal the Text attributed.string

https://github.com/fabriziobertoglio1987/react-native/blob/119892c46afb5456d8c374a73c2e0343ab91e979/React/Fabric/Mounting/ComponentViews/Text/RCTParagraphComponentView.mm#L150-L153

<details><summary>Android - Text correctly announcing accessibilityLabel</summary>
<p>

<video src="https://user-images.githubusercontent.com/24992535/204052236-5bee4103-9166-49e5-bf76-5f46dead55b3.mp4" />

</p>
</details>


<details><summary>iOS - not announcing accessibilityLabel</summary>
<p>

<video src="https://user-images.githubusercontent.com/24992535/204052405-8bc2f592-0796-47d0-a9bf-517a5ccb277f.mp4" />

</p>
</details>
